[No QA] Add tech debt philosophy doc (#95768)
Expensify/App: contributingGuides/philosophies/TECH-DEBT.md

> # Tech Debt Philosophy

Reviewer: ```suggestion
# Technical Debt Philosophy
```

I think it reads and looks nicer to spell out full words

> This philosophy guides how we identify, prioritize, and remove tech debt so it's handled consistently.

Reviewer 2: ```suggestion
This philosophy guides how we identify, prioritize, and remove technical debt so it's handled consistently.
```

#### Related Philosophies
- [Over-engineering Philosophy](/contributingGuides/philosophies/OVERENGINEERING.md)
- [Small Incremental Releases Philosophy](/contributingGuides/philosophies/INCREMENTAL-RELEASES.md)

#### Terminology

> - **Tech debt** - Code, architecture, or process shortcuts taken to ship faster or stale and outdated code that create extra cost (bugs, slower delivery, harder onboarding) if left unaddressed.

Reviewer: NAB - they're not necessarily always shortcuts, sometimes they're a necessity to shipping things smoothly. A good recent example is `isPolicyExpenseChatEnabled` (see https://github.com/Expensify/Expensify/issues/370320)

> - **Tech debt** - Code, architecture, or process shortcuts taken to ship faster or stale and outdated code that create extra cost (bugs, slower delivery, harder onboarding) if left unaddressed.

Reviewer 2: ```suggestion
- **Technical debt** - Code, architecture, or process shortcuts taken to ship faster or stale and outdated code that create extra cost (bugs, slower delivery, harder onboarding) if left unaddressed.
```

## Rules

### Tech debt MUST be reported as a GitHub issue, not a TODO comment

> Do not leave `TODO` comments in the code as a way to track debt. Instead, open a GH issue describing the problem and its impact, and reference it from a code comment if useful context is needed at the call site.

Reviewer 3: I think we already have something flagging TODOs in code in the AI reviewer, no?

Author: Not sure. Regardless, I don't think it'd be bad to add that here

> ### Example
> ❌ BAD: `// TODO: refactor this once we have time`
> ✅ GOOD: A GH issue describing the work that needs to be done, linked from a short comment if needed

Reviewer: There's a line break missing here
```suggestion
❌ BAD: `// TODO: refactor this once we have time`

✅ GOOD: A GH issue describing the work that needs to be done, linked from a short comment if needed
```

### Tech debt introduced to unblock a deploy or fire MUST be fixed promptly

> If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately and prioritized so it doesn't get forgotten and becomes a bad pattern to be followed.

Reviewer: NAB: This doesn't really say how to prioritize it, nor what the consequences are when we don't.

> If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately and prioritized so it doesn't get forgotten and becomes a bad pattern to be followed.

Reviewer 2: ```suggestion
If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately, assigned, and labeled Daily and FireCleanup so it's addresses with urgency.
```

Reviewer: I don't agree that technical debt issues should be Daily Fire-Cleanup

Author: I think that's because this is under the `### Technical debt introduced to unblock a deploy or fire MUST be fixed promptly` section

> ### Fixing tech debt SHOULD be scoped independently from feature work

Reviewer 2: ```suggestion
### Fixing technical debt SHOULD be scoped independently from feature work
```

Do not bundle a tech debt fix into an unrelated feature PR. Bundling makes the PR harder to review and harder to revert, and it obscures the cost/benefit of the cleanup itself. Work on the fix as its own PR.

### Example
❌ BAD: Refactoring a shared component while implementing a new, unrelated feature in the same PR

> ✅ GOOD: Refactoring the tech debt in one PR and then shipping the feature on a follow-up PR.

Reviewer: ```suggestion
❌ BAD: Refactoring a shared component while implementing a new, unrelated feature in the same PR

✅ GOOD: Refactoring the tech debt in one PR and then shipping the feature on a follow-up PR.
```

# Technical Debt Philosophy
This philosophy guides how we identify, prioritize, and remove technical debt so it's handled consistently.

#### Related Philosophies

> - [Over-engineering Philosophy](/contributingGuides/philosophies/OVERENGINEERING.md)

Reviewer 4: NAB all parts of hyphenated words are capitalized in title case

```suggestion
- [Over-Engineering Philosophy](/contributingGuides/philosophies/OVERENGINEERING.md)
```

### Example
❌ BAD: `// TODO: refactor this once we have time`

✅ GOOD: A GH issue describing the work that needs to be done, linked from a short comment if needed

### Technical debt introduced to unblock a deploy or fire MUST be fixed promptly
If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately and assigned an owner.

### How to prioritize it
Match the priority label to the risk the technical debt introduces, not just how inconvenient it is to fix:
- `Daily`/`Weekly` - the debt touches high-traffic or fragile code, or could mask a regression.
- `Monthly` - the debt is low-risk (e.g. a naming inconsistency or an isolated one-off).

### Consequences of not prioritizing it
- The technical debt is forgotten and becomes permanent.
- Other engineers/agents copy it elsewhere, treating it as an accepted pattern.
- The original context is lost over time, making the eventual fix more expensive and riskier than fixing it promptly would have been.

### Fixing technical debt SHOULD be scoped independently from feature work

> Do not bundle a technical debt fix into an unrelated feature PR. Bundling makes the PR harder to review and harder to revert, and it obscures the cost/benefit of the cleanup itself. Work on the fix as its own PR.

Reviewer 4: NAB I'm not really sure what "obscures the cost/benefit of the cleanup" means

```suggestion
Do not bundle a technical debt fix into an unrelated feature PR. Bundling makes the PR harder to review and harder to revert, and it's easier to miss subtle or unintentional business logic changes. Work on the fix as its own PR.
```

# Technical Debt Philosophy
This philosophy guides how we identify, prioritize, and remove technical debt so it's handled consistently.

#### Related Philosophies
- [Over-Engineering Philosophy](/contributingGuides/philosophies/OVERENGINEERING.md)
- [Small Incremental Releases Philosophy](/contributingGuides/philosophies/INCREMENTAL-RELEASES.md)

#### Terminology
- **Technical debt** - Code, architecture, or process decisions - whether shortcuts taken to ship faster or necessary tradeoffs made to ship smoothly - that become stale or outdated and create extra cost (bugs, slower delivery, harder onboarding) if left unaddressed.

## Rules

### Technical debt MUST be reported as a GitHub issue, not a TODO comment
Do not leave `TODO` comments in the code as a way to track debt. Instead, open a GH issue describing the problem and its impact, and reference it from a code comment if useful context is needed at the call site.

### Example
❌ BAD: `// TODO: refactor this once we have time`

✅ GOOD: A GH issue describing the work that needs to be done, linked from a short comment if needed

### Technical debt introduced to unblock a deploy or fire MUST be fixed promptly

> If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately, assigned, and labeled Daily and FireCleanup so it's addresses with urgency.

Reviewer: ```suggestion
If a shortcut is taken under time pressure (e.g., a deploy blocker fix), the follow-up GH issue MUST be filed immediately, assigned, and labeled Daily and Fire-Cleanup so it's addressed with urgency.
```

### Example
❌ BAD: `// TODO: refactor this once we have time`

✅ GOOD: A GH issue describing the work that needs to be done, linked from a short comment if needed

### Technical debt introduced to unblock a deploy or fire MUST be fixed promptly
If a shortcut is taken under time pressure (e.g. a deploy blocker fix), the follow-up GH issue MUST be filed immediately and assigned an owner.

### How to prioritize it
Match the priority label to the risk the technical debt introduces, not just how inconvenient it is to fix:
- `Daily`/`Weekly` - the debt touches high-traffic or fragile code, or could mask a regression.
- `Monthly` - the debt is low-risk (e.g. a naming inconsistency or an isolated one-off).

### Consequences of not prioritizing it
- The technical debt is forgotten and becomes permanent.
- Other engineers/agents copy it elsewhere, treating it as an accepted pattern.
- The original context is lost over time, making the eventual fix more expensive and riskier than fixing it promptly would have been.

> 

Reviewer: - The code becomes harder to read and reason about. 

Optionally, also, something like:
- This is like a death by a thousand cuts. A single small piece of isolated technical debt isn't so bad. If we have a thousand of them sprinkled throughout the codebase, the code's quality gets worse.

Reviewer 4: > * The code becomes harder to read and reason about.

👍 

> * This is like a death by a thousand cuts [...]

I'm not sure this one adds much that's not already written, though I do agree
